qualtran/bloqs/arithmetic.py

```diff
@@ -50,10 +54,54 @@ def t_complexity(self):
         num_t_gates = 4 * self.bitsize - 4
         return TComplexity(t=num_t_gates, clifford=num_clifford)
 
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set[Tuple[int, Bloq]]:
+        num_clifford = (self.bitsize - 2) * 19 + 16
+        num_t_gates = 4 * self.bitsize - 4
+        return {(num_t_gates, TGate()), (num_clifford, ArbitraryClifford(n=1))}
+
+
+@frozen
+class OutOfPlaceAdder(Bloq):
+    r"""An n-bit addition gate.
+
+    Implements $U|a\rangle|b\rangle 0\rangle \rightarrow |a\rangle|b\rangle|a+b\rangle$ using $4n - 4 T$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer. Must be large
+            enough to hold the result in the output register of a + b.
+
+    Registers:
+     - a: A bitsize-sized input register (register a above).
+     - b: A bitsize-sized input/output register (register b above).
+
+    References:
+        [Halving the cost of quantum addition](https://arxiv.org/abs/1709.06648)
+    """
+
+    bitsize: int
+
+    @property
+    def signature(self):
+        return Signature.build(a=self.bitsize, b=self.bitsize, c=self.bitsize)
+
+    def pretty_name(self) -> str:
+        return "c = a + b"
+
+    # TODO: This is just copied from above, not sure if correct.
+    def t_complexity(self):
+        num_clifford = (self.bitsize - 2) * 19 + 16
+        num_t_gates = 4 * self.bitsize - 4
+        return TComplexity(t=num_t_gates, clifford=num_clifford)
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set[Tuple[int, Bloq]]:
+        num_clifford = (self.bitsize - 2) * 19 + 16
+        num_t_gates = 4 * self.bitsize - 4
+        return {(num_t_gates, TGate()), (num_clifford, ArbitraryClifford(n=1))}
+
 
 @frozen
 class Square(Bloq):
-    r"""Square an n-bit number.
+    r"""Square an n-bit binary number.
```

Comment:
why is it relevant that they are binary? anything with a bitsize > 1 is treated like an opaque uint type where one cannot access individual bits

Reply:
There is mostly to distinguish integer squaring here with squaring floats later. Integer is probably better? The theorems in the ref start with "The complexity of squaring an n-bit binary number..." too.

Reply:
yeah integer would probably be better to my naive understanding, but if that's what the reference says then I leave it up to your judgement